### ISSUE TITLE
'Unknown error code 104 details Connection reset by peer' when use https to External services

### ISSUE THREAD

**Issue by pantianying** [istio/istio#36377]
### Bug Description

this is my config：
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: prod-kubebox
  namespace:dna
spec:
  hosts:
    - prod-kubebox.xiaodiankeji.net
  location: MESH_EXTERNAL
  ports:
    - name: https-443
      number: 443
      protocol: HTTPS
  resolution: DNS
```
```
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: prod-kubebox.xiaodiankeji.net
  namespace: dna
spec:
  host: prod-kubebox.xiaodiankeji.net
  trafficPolicy:
    tls:
      mode: SIMPLE
```
when I use no envoy sidecar, it is ok:
```
sh-4.4$ curl -I https://prod-kubebox.xiaodiankeji.net 
HTTP/1.1 200 OK
server: nginx/1.13.10
date: Sat, 04 Dec 2021 08:13:34 GMT
content-type: text/html; charset=UTF-8
content-length: 3627
last-modified: Fri, 03 Dec 2021 10:50:59 GMT
etag: "61a9f693-e2b"
accept-ranges: bytes
set-cookie: 977b88012255561c286781cd560d98b7=08d1f63576e5dbeabfc0b76266a57d4e; path=/; HttpOnly; Secure; SameSite=None
cache-control: private
```
when use istio-proxy sidecar
```
sh-4.4$ curl -I https://prod-kubebox.xiaodiankeji.net   
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
```
when I debug envoy log,  envoy had made a connection to the correct IP but failed for some reason:
```
2021-12-04T06:59:04.208830Z debug envoy filter  original_dst: New connection accepted
2021-12-04T06:59:04.208884Z debug envoy filter  tls inspector: new connection accepted
2021-12-04T06:59:04.218346Z debug envoy filter  tls:onServerName(), requestedServerName: prod-kubebox.xiaodiankeji.net
2021-12-04T06:59:04.218466Z debug envoy filter  [C173669] new tcp proxy session
2021-12-04T06:59:04.218490Z debug envoy filter  [C173669] Creating connection to cluster outbound|443||prod-kubebox.xiaodiankeji.net
2021-12-04T06:59:04.218514Z debug envoy pool  creating a new connection
2021-12-04T06:59:04.218575Z debug envoy pool  [C173670] connecting
2021-12-04T06:59:04.218580Z debug envoy connection  [C173670] connecting to 10.111.255.222:443
2021-12-04T06:59:04.219593Z debug envoy connection  [C173670] connection in progress
2021-12-04T06:59:04.219945Z debug envoy pool  queueing request due to no available connections
2021-12-04T06:59:04.219987Z debug envoy conn_handler  [C173669] new connection
2021-12-04T06:59:04.222165Z debug envoy connection  [C173670] connected
2021-12-04T06:59:04.227731Z debug envoy pool  [C173670] assigning connection
2021-12-04T06:59:04.227915Z debug envoy filter  [C173669] TCP:onUpstreamEvent(), requestedServerName: prod-kubebox.xiaodiankeji.net
2021-12-04T06:59:04.231807Z debug envoy misc  Unknown error code 104 details Connection reset by peer
2021-12-04T06:59:04.232116Z debug envoy connection  [C173669] remote close
2021-12-04T06:59:04.232170Z debug envoy connection  [C173669] closing socket: 0
2021-12-04T06:59:04.232315Z debug envoy wasm  expr_evaluate value error: No value with name "upstream_peer" found in Activation
2021-12-04T06:59:04.232369Z debug envoy wasm  wasm log: [extensions/stats/plugin.cc:618]::report() metricKey cache hit , stat=28
2021-12-04T06:59:04.232376Z debug envoy wasm  wasm log: [extensions/stats/plugin.cc:618]::report() metricKey cache hit , stat=32
2021-12-04T06:59:04.232379Z debug envoy wasm  wasm log: [extensions/stats/plugin.cc:618]::report() metricKey cache hit , stat=36
2021-12-04T06:59:04.232382Z debug envoy wasm  wasm log: [extensions/stats/plugin.cc:618]::report() metricKey cache hit , stat=40
2021-12-04T06:59:04.232389Z debug envoy connection  [C173670] closing data_to_write=0 type=0
2021-12-04T06:59:04.232391Z debug envoy connection  [C173670] closing socket: 1
2021-12-04T06:59:04.232428Z debug envoy connection  [C173670] SSL shutdown: rc=1
2021-12-04T06:59:04.232437Z debug envoy connection  [C173670]
2021-12-04T06:59:04.232459Z debug envoy pool  [C173670] client disconnected
2021-12-04T06:59:04.232470Z debug envoy conn_handler  [C173669] adding to cleanup list
2021-12-04T06:59:04.232478Z debug envoy pool  [C173670] connection destroyed
[2021-12-04T06:59:04.208Z] "- - -" 0 - - - "-" 517 207 23 - "-" "-" "-" "-" "10.111.255.222:443" outbound|443||prod-kubebox.xiaodiankeji.net 10.140.14.111:34966 10.111.255.222:443 10.140.14.111:34964 prod-kubebox.xiaodiankeji.net -
```
some trace log

```
2021-12-04T08:36:29.195989Z	trace	envoy connection	[C2770] read ready. dispatch_buffered_data=false
2021-12-04T08:36:29.195998Z	trace	envoy connection	[C2770] ssl read returns: -1
2021-12-04T08:36:29.196002Z	trace	envoy connection	[C2770] ssl read 0 bytes
2021-12-04T08:36:29.196009Z	trace	envoy connection	[C2769] socket event: 3
2021-12-04T08:36:29.196010Z	trace	envoy connection	[C2769] write ready
2021-12-04T08:36:29.196013Z	trace	envoy connection	[C2769] read ready. dispatch_buffered_data=false
2021-12-04T08:36:29.196021Z	trace	envoy connection	[C2769] read returns: 517
2021-12-04T08:36:29.196028Z	trace	envoy connection	[C2769] read error: Resource temporarily unavailable
2021-12-04T08:36:29.196036Z	trace	envoy filter	[C2769] downstream connection received 517 bytes, end_stream=false
2021-12-04T08:36:29.196041Z	trace	envoy filter	Alpn Protocol Not Found. Expected istio-peer-exchange, Got
2021-12-04T08:36:29.196048Z	trace	envoy connection	[C2770] writing 517 bytes, end_stream false
2021-12-04T08:36:29.196055Z	trace	envoy connection	[C2770] socket event: 2
2021-12-04T08:36:29.196056Z	trace	envoy connection	[C2770] write ready
2021-12-04T08:36:29.196101Z	trace	envoy connection	[C2770] ssl write returns: 517
2021-12-04T08:36:29.197382Z	trace	envoy connection	[C2770] socket event: 3
2021-12-04T08:36:29.197394Z	trace	envoy connection	[C2770] write ready
2021-12-04T08:36:29.197399Z	trace	envoy connection	[C2770] read ready. dispatch_buffered_data=false
2021-12-04T08:36:29.197418Z	trace	envoy connection	[C2770] ssl read returns: 207
2021-12-04T08:36:29.197446Z	trace	envoy connection	[C2770] ssl read returns: 0
2021-12-04T08:36:29.197451Z	trace	envoy connection	[C2770] ssl read 207 bytes
2021-12-04T08:36:29.197462Z	trace	envoy filter	[C2769] upstream connection received 207 bytes, end_stream=true
2021-12-04T08:36:29.197470Z	trace	envoy connection	[C2769] writing 207 bytes, end_stream true
2021-12-04T08:36:29.197479Z	trace	envoy connection	[C2770] socket event: 3
2021-12-04T08:36:29.197483Z	trace	envoy connection	[C2770] write ready
2021-12-04T08:36:29.197485Z	trace	envoy connection	[C2770] read ready. dispatch_buffered_data=false
2021-12-04T08:36:29.197488Z	trace	envoy connection	[C2770] ssl read returns: 0
2021-12-04T08:36:29.197490Z	trace	envoy connection	[C2770] ssl read 0 bytes
2021-12-04T08:36:29.197492Z	trace	envoy connection	[C2769] socket event: 2
2021-12-04T08:36:29.197494Z	trace	envoy connection	[C2769] write ready
2021-12-04T08:36:29.197529Z	trace	envoy connection	[C2769] write returns: 207
2021-12-04T08:36:29.197547Z	trace	envoy connection	[C2769] socket event: 2
2021-12-04T08:36:29.197549Z	trace	envoy connection	[C2769] write ready
2021-12-04T08:36:29.198157Z	trace	envoy connection	[C2769] socket event: 3
2021-12-04T08:36:29.198160Z	trace	envoy connection	[C2769] write ready
2021-12-04T08:36:29.198162Z	trace	envoy connection	[C2769] read ready. dispatch_buffered_data=false
2021-12-04T08:36:29.198170Z	trace	envoy connection	[C2769] read error: Connection reset by peer
```

### Version

```prose
data plane version: 1.9.8 (12 proxies)
```

### Additional Information

You may ask why I need to configure the destinationRule configuration, it's fine when it's not configured. But I need the gateway to route certain https requests to the service configured as follows:
```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: test
  namespace: dna
spec:
  gateways:
    - dna/dna-gateway
  hosts:
    - hub-dna.xiaodiankeji.net
  http:
    - match:
        - uri:
            prefix: /prod-kubebox
      name: prod-kubebox
      route:
        - destination:
            host: prod-kubebox.xiaodiankeji.net
            port:
              number: 443
          headers:
            request:
              set:
                host: prod-kubebox.xiaodiankeji.net
```
If I do not configure destinationRule above then this ingress traffic cannot be sent to prod-kubebox.xiaodiankeji.net using HTTPS. 

To sum up, I have encountered a complex configuration problem, I do not know whether it is a bug or there is a problem with my configuration, please help community members to answer it.

**Comment by bianpengyuan** [comment]
Can you move the gateway virtual service and destination rule to `dna` namespace?

**Comment by pantianying** [comment]
> Can you move the gateway virtual service and destination rule to `dna` namespace?

Sorry, in fact, all the above configurations are under the namspace of `dna`

**Comment by bianpengyuan** [comment]
Probably update your app to make plain text http call instead and let sidecar proxy handle TLS initiation?

**Comment by pantianying** [comment]
> Probably update your app to make plain text http call instead and let sidecar proxy handle TLS initiation?

 I have tested that this and it does not work, why not keep TLS traffic transparent to external services

**Comment by bianpengyuan** [comment]
> I have tested that this and it does not work

You will need to update service entry port protocol to http in order to make it accept plain text traffic.

> why not keep TLS traffic transparent to external services

Because you are configuring proxies in `dna` namespace to initiate TLS when talking to the external service, and the external service can only terminate TLS once. In this case your app is also initiating TLS in addition to the sidecar proxy, which does not work. Changing app is indeed not ideal, the ways I can think of to work around this are 1) define a different subset for gateway TLS initiation, or 2) move gateway to a different namespace.

**Comment by pantianying** [comment]
> > I have tested that this and it does not work
> 
> You will need to update service entry port protocol to http in order to make it accept plain text traffic.
> 
> > why not keep TLS traffic transparent to external services
> 
> Because you are configuring proxies in `dna` namespace to initiate TLS when talking to the external service, and the external service can only terminate TLS once. In this case your app is also initiating TLS in addition to the sidecar proxy, which does not work. Changing app is indeed not ideal, the ways I can think of to work around this are 1) define a different subset for gateway TLS initiation, or 2) move gateway to a different namespace.

ths for reply, but one I don't understand is 
'define a different subset for gateway TLS initiation'
Can you explain what this configuration is? Define another destinationRule? ths~

**Comment by bianpengyuan** [comment]
Something like

```
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: prod-kubebox.xiaodiankeji.net
  namespace: dna
spec:
  host: prod-kubebox.xiaodiankeji.net
  subsets:
  - name: gateway
    trafficPolicy:
      tls:
        mode: SIMPLE
```

then at your gateway virtual service, route matched traffic to the gateway subset:

```
      route:
        - destination:
            host: prod-kubebox.xiaodiankeji.net
            subset: gateway
            port:
              number: 443
```

**Comment by pantianying** [comment]
> Something like
> 
> ```
> apiVersion: networking.istio.io/v1beta1
> kind: DestinationRule
> metadata:
>   name: prod-kubebox.xiaodiankeji.net
>   namespace: dna
> spec:
>   host: prod-kubebox.xiaodiankeji.net
>   subsets:
>   - name: gateway
>     trafficPolicy:
>       tls:
>         mode: SIMPLE
> ```
> 
> then at your gateway virtual service, route matched traffic to the gateway subset:
> 
> ```
>       route:
>         - destination:
>             host: prod-kubebox.xiaodiankeji.net
>             subset: gateway
>             port:
>               number: 443
> ```

Thank you for your guidance, which has solved my current problem. I will close the issue
